File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        # combs = []
        nums = candidates

#         def backtrack(i, curComb, curSum):
#             if curSum == target:
#                 combs.append(curComb.copy())
#                 return
            
#             if i >= len(nums) or curSum > target:
#                 return
            
#             # choose num between 1 and unlimited times
#             curComb.append(nums[i])
#             backtrack(i, curComb, curSum + nums[i])
            
#             # skip num
#             curComb.pop()
#             backtrack(i+1, curComb, curSum)
        
#         backtrack(0, [], 0)
#         return combs
# Time: O(2^T*N) where T is the target value and N is the number of candidates
# Space: O(T) where T is the target value
        
        combs = set()
        def backtrack(i, cur_comb, cur_sum):
            if cur_sum > target:
                return
            if cur_sum == target:
                combs.add(tuple(sorted(cur_comb)))
            for j in range(i, len(nums)):
                cur_comb.append(nums[j])
                backtrack(j, cur_comb, cur_sum+ nums[j])
                cur_comb.pop()
        backtrack(0,[], 0)
        return list(combs)
        
        

File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const ans = [];
    function backtrack(index, path, curSum) {
        if (curSum === target) {
            ans.push([...path])
            return;
        }

        if (curSum > target) {
            return;
        }

        for (let i = index; i < candidates.length; i++) {
            path.push(candidates[i]);
            backtrack(i, path, curSum + candidates[i]);
            path.pop();
        }
    }

    backtrack(0, [], 0)
    return ans;





};